### ISSUE TITLE
objective function modification

### ISSUE THREAD

**Issue by zhen0z** [sintefmath/JutulDarcy.jl#100]
Hi Olav,

I'm trying to use this objective function:

function objective_function(model, state, Δt, step_i, forces)
        sg = state.Reservoir.Saturations[2, :]
        loss = 0.0
        # Compute loss if step_i is in observation steps
        if step_i in steps
            loss = 0.5 * sum(sg  .- observation[step_i]) .^ 2)
        end
        return loss
    end

It turns out the computational time is large if I compute the loss step by step, especifically if I do some transformation of the saturation into other properties. Is it possible to compute the loss in a single run instead step by step? Thanks.

Best regards,
Zhen

**Comment by moyner** [comment]
At the moment the code assumes that the objective function is a sum. Part of this is the adjoint method JutulDarcy uses for the gradients - even if we implemented the objective function as a single function for all steps it would have to be called once per step.

I think part of the reason is that you are implicitly copying the states several times. Can you try something like the following:
```julia
function objective_function(model, state, Δt, step_i, forces)
    S = Saturations
    loss = zero(eltype(S))
    # Compute loss if step_i is in observation steps
    if step_i in steps
        for cell in axes(sg, 2)
            loss += 0.5 * (S[2, cell] - observation[step_i, cell]) ^ 2
        end
    end
    return loss
end
```
You can also consider to do `use_sparsity = false` in your call to `setup_parameter_optimization` since your objective function is very dense. I think reformulating the function to avoid making copies of the saturations is probably more important though. Let me know if this makes sense...

**Comment by zhen0z** [comment]
Thanks! This was the cause of the issue—now the code runs significantly faster.

Best regards,
Zhen

**Comment by zhen0z** [comment]
Could you clarify what do you mean for S = Saturations?

I think I misunderstand your point, if I do this, the computational time is not improving a lot:

function objective_function(model, state, Δt, step_i, forces)
    S = state.Reservoir.Saturations
    loss = 0.0
    # Compute loss if step_i is in observation steps
    if step_i in steps
        loss = 0.5 * sum(S[2, :] .- observation[step_i]) .^ 2)
    end
    return loss
end

If I do this (use the states from the previous simulation), the computational time is less but the gradient is not computed (zero everywhere):

result = simulate_reservoir(state0, model, dt, parameters = parameters, forces = forces)
ws, states = result
Saturations = [(states[step][:Saturations][2, :]) for step in allsteps]
function objective_function(model, state, Δt, step_i, forces)
    S = Saturations
    loss = 0.0
    # Compute loss if step_i is in observation steps
    if step_i in steps
        loss = 0.5 * sum(S[step_i][2, :] .- observation[step_i]) .^ 2)
    end
    return loss
end

Thanks.

**Comment by moyner** [comment]
Here is an updated version. I was missing the access for the saturations. The main thing to consider is that slicing (e.g. `S[step_i][2, :]`) like you have in your revised function will create a copy of a large part of the saturations matrix, will will be very slow. As long as you avoid `:` it should speed things up.

```julia
function objective_function(model, state, Δt, step_i, forces)
    S = state[:Reservoir][:Saturations]
    loss = zero(eltype(S))
    # Compute loss if step_i is in observation steps
    if step_i in steps
        for cell in axes(sg, 2) # Note the loop to avoid making intermediate vectors
            loss += 0.5 * (S[2, cell] - observation[step_i][cell]) ^ 2
        end
    end
    return loss
end
```
If you have a short example that demonstrates the issue I can have a look - usually it is a question of making sure that the objective function does not create any big objects when evaluated since this happens many times: In your case it will be called once per cell, per timestep (potentially millions of times if you have 10,000+ cells with 1000 time-steps)

**Comment by zhen0z** [comment]
I think I need some guidance here:

Let's say I have a permeability field with dimension of 20 by 20 by 1, I run it for 10 years with 100 time steps, then I want to extract one saturation map in each year as observation data (10 saturation map as observation data), here's what I do:

First, run simulation:
'''
result = simulate_reservoir(state0, model, dt, parameters = parameters, forces = forces)
ws, states = result
'''

Second, extract 10 saturation maps in 10 time steps:
'''
steps = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
obs = reduce(vcat, [vec(states[step][:Saturations][2, :]) for step in steps])
'''

Third, use the observation data obs to compute gradient:
'''
steps = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
datasize = 20 * 20 * 1            
function objective_function(model, state, Δt, step_i, forces)
    loss = 0.0
    # Compute loss if step_i is in observation steps
    if step_i in steps
        obs = data[1 + div(step_i,10) * datasize : (div(step_i,10)+1) * datasize]        
        loss = 0.5 * sum((state.Reservoir.Saturations[2, :] .- obs) .^ 2)
    end
'''
(datasize is used to extract one saturation map from the obs data)
( div(step_i,10): because in time step 10, 20,...,100   we need to extract the corresponding saturation data stored in obs )

Is it possible to accelerate the code? I don't understand why should we use: for cell in axes(sg, 2), because I need to use all the saturation data in all the permeability cells at specific time steps...
Thanks.

**Comment by zhen0z** [comment]
I think I need some guidance here:

Let's say I have a permeability field with dimension of 20 by 20 by 1, I run it for 10 years with 100 time steps, then I want to extract one saturation map in each year as observation data (10 saturation map as observation data), here's what I do:

First, run simulation:
```
result = simulate_reservoir(state0, model, dt, parameters = parameters, forces = forces)
ws, states = result
```

Second, extract 10 saturation maps in 10 time steps:
```
steps = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
obs = reduce(vcat, [vec(states[step][:Saturations][2, :]) for step in steps])
```

Third, use the observation data obs to compute gradient:
```
steps = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100]
datasize = 20 * 20 * 1
function objective_function(model, state, Δt, step_i, forces)
    loss = 0.0
    # Compute loss if step_i is in observation steps
    if step_i in steps
        obs = data[1 + div(step_i,10) * datasize : (div(step_i,10)+1) * datasize]
        loss = 0.5 * sum((state.Reservoir.Saturations[2, :] .- obs) .^ 2)
    end
    return loss
end
```
(datasize is used to extract one saturation map from the obs data)
( div(step_i,10): because in time step 10, 20,...,100 we need to extract the corresponding saturation data stored in obs )

Is it possible to accelerate the code? I don't understand why should we use: for cell in axes(sg, 2), because I need to use all the saturation data in all the permeability cells at specific time steps...
Thanks.

**Comment by moyner** [comment]
Hi,

I think your approach makes sense to me. `axes(sg, 2)` will expand to `1:size(sg, 2)` which means that the loop will go over all cells. The code I posted should give identical results to your posted objective function, but will be much, much faster to run. The idea is that calling `state.Reservoir.Saturations[2, :]` will make a (expensive) copy of all saturations for one phase while doing a loop will be inexpensive. Here is a small example to demonstrate:
```julia
using Test

function sum_version1(A)
    x = zero(eltype(A))
    for i in axes(A, 2)
        x += A[2, i]
    end
    return x
end

function sum_version2(A)
    return sum(A[2, :])
end

function timing(N)
    A = rand(2, N)
    @test sum_version1(A) ≈ sum_version2(A)
    println("Version 1:")
    @time sum_version1(A)
    println("Version 2:")
    @time sum_version2(A)
end

# Precompile
timing(1)
# Actually test
timing(1_000_000)
```

At least for me, the last version is a lot faster and crucially allocates much less memory. The effect is even larger for a big matrix with AD numbers, which is what will happen during your simulation.

**Comment by zhen0z** [comment]
Thanks for the explanations, the computation is faster now. However, when I do this:
```
    function objective_function(model, state, Δt, step_i, forces)
        loss = 0.0
        # Compute loss if step_i is in observation steps
        if step_i in steps
            sg = state.Reservoir.Saturations
            for cell in axes(sg, 2)
                velocity = sqrt(((1.0 / ((1.0 - sg[2,cell]) / (bulk_sat1 + term4) + sg[2,cell] / 
                ((bulk_min / (1.0 / (term1 - term2 / ϕ[cell] + term3 / ϕ[cell]) + 1.0)) + term4)) - term4) + term4) / 
                (rhoR * (1 - ϕ[cell]) + ϕ[cell] * rhoL + ϕ[cell] * sg[2,cell] * rhodif))
                loss += 0.5 * (velocity - data[cell]) ^ 2
            end
        end
        return loss
    end 
```
It is still a bit slow, is it possible to optimize it further?

**Comment by moyner** [comment]
Try the following:
```julia
function step_loss(sg)
  for cell in axes(sg, 2)
      velocity = sqrt(((1.0 / ((1.0 - sg[2,cell]) / (bulk_sat1 + term4) + sg[2,cell] / 
      ((bulk_min / (1.0 / (term1 - term2 / ϕ[cell] + term3 / ϕ[cell]) + 1.0)) + term4)) - term4) + term4) / 
      (rhoR * (1 - ϕ[cell]) + ϕ[cell] * rhoL + ϕ[cell] * sg[2,cell] * rhodif))
      loss += 0.5 * (velocity - data[cell]) ^ 2
  end
end

function objective_function(model, state, Δt, step_i, forces)
    loss = 0.0
    # Compute loss if step_i is in observation steps
    if step_i in steps
        sg = state.Reservoir.Saturations
        loss += step_loss(sg)
    end
    return loss
end
```
Putting the loop into a function closure could help with type inference which can sometimes be a cause of slowdown if you are evaluating a function many types. You can also check what the types of the global variables you are using have: https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-untyped-global-variables

(Global variables here would be bulk_sat1, term4, term3, etc)

**Comment by zhen0z** [comment]
Thanks for the guidance, now the code is accelerated more than 10 times. The computational time is fine now.
